### PR TITLE
[Misc] Add standalone safetensors retention diagnostic

### LINKED ISSUE #7143: [RFC] Isolate safetensors memory retention with a CPU diagnostic
### Motivation

Part of #7107. The tracker reports approximately 31 MiB of retained private memory after one million direct safetensors `get_tensor()` calls, including outside HWR. This does not establish per-request HWR leakage: HWR materializes views during lease acquisition and exposes a stable tensor mapping afterward.

Historical upstream report https://github.com/safetensors/safetensors/issues/369 describes retention with safetensors 0.3.1. It is not evidence of a confirmed cause or fix for the current dependency version.

### Goals

- Preserve a small, model-independent CPU diagnostic reproducing direct dependency calls separately from reuse of a cached view.
- Record versions, operation count, post-GC private memory, anonymous memory, RSS, descriptors, and elapsed time in machine-readable output.
- Distinguish live references, process allocator retention and file-backed residency; support repeating the same probe in separate environments for upstream investigation.

### Non-goals

- Declaring a memory leak from RSS alone, attributing retention to HWR request processing, or changing dependencies without a verified fix.
- Replacing safetensors, trimming allocators, changing shared page caches, or adding forced garbage collection to serving.
- An inference benchmark or an automatic pass/fail memory threshold across platforms.

### Use cases

1. Reproduce the tracker observation with synthetic CPU tensors and no vLLM/HWR import.
2. Compare repeated `get_tensor()` calls against repeated access to one retained tensor view in fresh processes.
3. Re-run against another dependency version and send versioned raw evidence to the dependency maintainer.

### Detailed design

Add a Linux-only diagnostic under `benchmarks/host_weight_runtime/`, with a README linked from HWR documentation. Each invocation runs exactly one mode in its own process. Create a synthetic float32 safetensors payload in a task-owned temporary directory. Expose a positive `--tensor-elements` parameter (default 64), record it in JSON, and use it for both file creation and the correctness check. Import dependencies, create/open the file and warm up before recording the baseline. Separate preparation from loop time.

`get_tensor` calls the dependency once per iteration and drops the returned tensor; `reuse` accesses one cached tensor. Record checkpoints after explicit GC, including `/proc/self/smaps_rollup` memory counters and open-FD count. Keep a weak reference to the last transient tensor to verify the script does not retain it. Close the reader, delete cached views, collect and record a final sample. Emit JSON containing arguments, software versions, host/process context, and every sample; caller supplies the repository SHA/dirty state with the report.

The fix design is diagnostic coverage and accurate operational documentation. A production dependency change requires a separate reproducer-backed upstream fix; this PR will not manufacture one.

### Experiment contract and validation

Hypothesis: repeated dependency materialization can retain private memory depending on tensor size; cached reuse avoids repeated materialization. Primary isolated variable: 16 versus 64 float32 elements, compared within each mode; cached reuse is a separate control at both sizes. Fixed controls: one CPU, one thread, same dependency environment, operation count, warmup, checkpoint cadence, and default allocator/cache state. Changing element count necessarily changes tensor shape and payload/header size. No accelerator, model download, shared-cache change or allocator trimming.

Run one short default-size feasibility probe, then two additional fresh-process measured repetitions per size/mode pair (eight measured runs), excluding feasibility. Success means valid complete samples, released transient references and closed descriptors, not a predetermined memory result. Stop on exception, 120 seconds per invocation, or a missing required Linux proc interface. Preserve exact commands and raw JSON. Static checks and CLI invalid-argument checks validate the diagnostic; no platform-dependent memory assertion belongs in CI.

Review follow-up: the reporter identified size dependence in #7145. Local runs reproduced 30.656–30.664 MiB of private-memory growth at 64 elements, while 16-element runs and cached reuse remained below 0.04 MiB. See the PR for both repetitions and the distinction between observed retention and an established root cause.

### PR DESCRIPTION
**Pull request by hsliuustc0106**
## Purpose

Fixes #7143 (diagnostic scope). Part of #7107.

Add a standalone Linux CPU diagnostic for safetensors tensor-materialization retention. The synthetic float32 tensor size is configurable with `--tensor-elements`, defaults to 64, and is recorded in JSON. Both file creation and the correctness check use the same value. This exposes size-dependent behavior that a fixed 16-element probe misses.

Compare repeated `get_tensor()` calls with reuse of one cached view, without importing vLLM or HWR. Record warmed baselines, private/anonymous memory, RSS, descriptors, released-object checks, versions and CPU affinity as JSON; separate preparation from loop timing. Document a 16-versus-64 size comparison with cached reuse as a control at both sizes.

HWR acquires views at lease startup and exposes cached tensors afterward, so this synthetic loop cannot establish per-request HWR leakage. No runtime or dependency behavior changes.

## Test Plan

Primary variable: 16 versus 64 float32 elements, compared within each mode. Controls: CPU 0 affinity, one Torch thread, default allocator, identical environment, 1,000 warmup operations, one million measured operations, samples every 100,000. No shared-cache manipulation or accelerator work. Run one 10,000-operation feasibility probe, then two additional fresh-process measured repetitions for each size/mode pair (eight runs); exclude feasibility from comparisons.

```bash
CUDA_VISIBLE_DEVICES='' taskset -c 0 timeout 120s \
  /tmp/codex-pr6607-vllm028/bin/python \
  benchmarks/host_weight_runtime/safetensors_retention.py \
  --mode get_tensor --tensor-elements 64
# Repeat twice per size (16, 64) and mode (get_tensor, reuse).
```

**Versions:** Python 3.12.13, torch 2.13.0+cu129, safetensors 0.8.0; Linux 5.10.134-16.103.al8.x86_64. PR base is vLLM-Omni `039808e0d97d7969a2cb102d074c1e45b8ceef0b`; validated code is `743953b821e5352f3958c781811d53c28702584b` (the frozen tensor-size patch). Script bytes and working diff were frozen across all eight runs; this probe does not import vLLM.

## Test Result

Private-memory growth from warmed baseline to final loop sample (MiB):

| Tensor elements | Mode | Run 1 | Run 2 |
| --- | --- | --- | --- |
| 16 | `get_tensor` | 0.031250 | 0.039062 |
| 16 | `reuse` | 0.027344 | 0.023438 |
| 64 | `get_tensor` | 30.664062 | 30.656250 |
| 64 | `reuse` | 0.023438 | 0.023438 |

- **The reported ~30.66 MiB growth reproduces with 64 elements.** The two runs differ by 8 KiB; 30.660–30.668 MiB remains after closing the reader and deleting views. Anonymous-memory deltas match private-memory deltas. The earlier 16-element negative result does not generalize to other tensor sizes.
- All transient/final tensor weak-reference checks passed, and descriptor counts remained at four. This narrows the observation to size-dependent dependency/allocator retention; it does not identify its root cause or prove per-request HWR leakage.
- Default 64-element feasibility, both sizes/modes and invalid tensor sizes (zero, negative, nonnumeric) were checked. All applicable local gates and full precheck pass. The mypy wrapper has no benchmark targets; no claim of benchmark type coverage.
- Raw JSON, exact commands, script hash and working diff are preserved locally with the task artifacts. No GPU, inference, throughput improvement, memory fix or automatic leak verdict is claimed.

### CODE REVIEW

File: benchmarks/host_weight_runtime/safetensors_retention.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    import torch
+    from safetensors import safe_open
+    from safetensors.torch import save_file
+
+    torch.set_num_threads(1)
+    torch.set_num_interop_threads(1)
+    samples = [sample("before_file", 0, 0.0)]
+    with tempfile.TemporaryDirectory(prefix="safetensors-retention-") as directory:
+        path = Path(directory) / "synthetic.safetensors"
+        save_file({"weight": torch.arange(16, dtype=torch.float32)}, str(path))
```
Comment (baonudesifeizhai) [inline benchmarks/host_weight_runtime/safetensors_retention.py]: I changed two occurrences of 16 to 64, leaving everything else unchanged, and measured an increase of approximately 30.66 MiB.

File: benchmarks/host_weight_runtime/safetensors_retention.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        with safe_open(path, framework="pt", device="cpu") as reader:
+            cached = reader.get_tensor("weight")
+            assert torch.equal(cached, torch.arange(16, dtype=torch.float32))
```
Comment (baonudesifeizhai) [inline benchmarks/host_weight_runtime/safetensors_retention.py]: and here
